sql/src/main/java/io/crate/executor/transport/AlterTableOperation.java

```diff
@@ -74,44 +78,90 @@
-                    Long count = (Long) sqlResponse.rows()[0][0];
-                    if (count == 0L) {
-                        addColumnToTable(analysis, result);
-                    } else {
-                        String columnFailure = analysis.newPrimaryKeys() ? "primary key" : "generated";
-                        result.setException(new UnsupportedOperationException(String.format(Locale.ENGLISH,
-                            "Cannot add a %s column to a table that isn't empty", columnFailure)));
-                    }
-                }
+            String schema = analysis.table().ident().schema();
+            String stmt = String.format(Locale.ENGLISH,
+                "SELECT COUNT(*) FROM \"%s\".\"%s\"", schema, analysis.table().ident().name()
+            );
 
-                @Override
-                public void onFailure(Throwable e) {
-                    result.setException(e);
-                }
-            });
+            SQLOperations.Session session = sqlOperations.createSession(schema, Option.NONE, 1);
+            try {
+                session.parse(UNNAMED, stmt, Collections.<DataType>emptyList());
+                session.bind(UNNAMED, UNNAMED, Collections.emptyList(), null);
+                session.execute(UNNAMED, 1, new ResultSetReceiver(analysis, result));
+                session.sync(CompletionListener.NO_OP);
+            } catch (Throwable t) {
+                result.setException(t);
+            }
         } else {
             addColumnToTable(analysis, result);
         }
         return result;
     }
 
+    private class ResultSetReceiver implements ResultReceiver {
+
+        private final AddColumnAnalyzedStatement analysis;
+        private final SettableFuture<Long> result;
+
+        private CompletionListener completionListener = CompletionListener.NO_OP;
+        private long count;
+
+        ResultSetReceiver(AddColumnAnalyzedStatement analysis, SettableFuture<Long> result) {
+            this.analysis = analysis;
+            this.result = result;
+        }
+
+        @Override
+        public void addListener(CompletionListener listener) {
+            this.completionListener = CompletionMultiListener.merge(this.completionListener, listener);
+        }
+
+        @Override
+        public void setNextRow(Row row) {
+            count = (long) row.get(0);
+        }
+
+        @Override
+        public void batchFinished() {
+        }
+
+        @Override
+        public void allFinished() {
+            if (count == 0L) {
+                addColumnToTable(analysis, result);
+            } else {
+                String columnFailure = analysis.newPrimaryKeys() ? "primary key" : "generated";
+                result.setException(new UnsupportedOperationException(String.format(Locale.ENGLISH,
+                    "Cannot add a %s column to a table that isn't empty", columnFailure)));
+            }
+            completionListener.onSuccess(null);
```

Comment:
completionListener is called with onSuccess even if `reset.setException` is called?